feat(images): accept url or base64 data

mckdanoventa opened the pull request:
Hi @wyozi , 

I know you want a more centralized solution for both background and image components. I'm using the contract of data that would be passed to the html src tag `data:<content type>;base64,<data>`. How's this approach for you? How would you like to see this implemented for approval? 

I tested it locally via preview and test generated ppt. Works well so far. 

## Generated PPT
### Test image crated with small data file
<img width="526" alt="Screen Shot 2020-11-20 at 2 47 33 PM" src="https://user-images.githubusercontent.com/73127396/99852910-935dc300-2b47-11eb-8302-0f2f39ca86e3.png">
<img width="1176" alt="Screen Shot 2020-11-20 at 2 47 38 PM" src="https://user-images.githubusercontent.com/73127396/99852913-948ef000-2b47-11eb-8ed2-2edcdb0f6626.png">

## Render 
### Preview code and replaced image
<img width="778" alt="Screen Shot 2020-11-20 at 3 35 59 PM" src="https://user-images.githubusercontent.com/73127396/99852917-9658b380-2b47-11eb-9126-ac9700e8c0cb.png">
<img width="923" alt="Screen Shot 2020-11-20 at 3 35 51 PM" src="https://user-images.githubusercontent.com/73127396/99852908-92c52c80-2b47-11eb-97ec-5040550aac89.png">

## Error when no url or data
How would you like to handle this? Only one prop from the image component? `url` with some base64 regex? 
<img width="473" alt="Screen Shot 2020-11-20 at 2 42 37 PM" src="https://user-images.githubusercontent.com/73127396/99852909-935dc300-2b47-11eb-8a56-193c9671c3b1.png">

wyozi commented:
Yeah, I'm not sure about using separate props for data and url. Is there ever a case where both would need to be specified (which would be the biggest reason to have the two imo)?

The idea of clearly separating between urls and data is solid, but even then it could even be done with a single prop e.g. `<Image src={{kind: "data", data: "..."}}>`. Now that specific example syntax is quite ugly, but it has some nice properties like only one prop name you need to remember, very clear error messages, and reasonable fallback to just passing in a string for a URL.

> url with some base64 regex?
This could be it. It's not the nicest or the most efficient solution, but there is (even if questionable) prior art in how HTML `<img>` works and it keeps the api reasonable slim.

mckdanoventa commented:
Agree. One prop makes sense, and I like the src tag approach you highlighted vs. using regex. 

Would you like to keep url? in place for the time being vs. introducing a breaking change with a deprecated console message? Or get the breaking change in now?

wyozi commented:
@mckdanoventa I think with library of this size early breaking changes are okay and good to keep the code clean especially if they're this easily fixable.

I'm wondering if the src prop should still accept a plain string as the URL? I think url would still make sense as the default case, although having two types of input props could cause some confusion.

mckdanoventa commented:
Hi @wyozi. I updated it to only support the `src` prop. I agree, I think having both supported might get a bit confusing, and would only use it as a stop gap. That being said, if you wanted to support both, we could implement a fall back to regex image vs. http prefix. and have string or type. 

I've updated the preview and test examples. Slide Background and Images are now supporting the `data` or `path` workflows.

wyozi commented on src/normalizer.ts:
```diff
@@ -76,10 +77,15 @@ export type InternalShape = ObjectBase & {
 
 export type InternalSlideObject = InternalText | InternalImage | InternalShape;
 
+export type InternalImageSrc = {
```
This could also be `{ kind: "data", data: string } | { kind: "path", path: string }` to make the attribute a bit nicer for path

mckdanoventa replied:
👍

wyozi commented on src/preview/Preview.tsx:
```diff
@@ -161,7 +162,9 @@ const SlideObjectPreview = ({
       ) : object.kind === "image" ? (
         <img
-          src={object.url}
+          src={object.src && object.src.kind === 'data'
```
Small nitpick 😁 (applies to the next two lines too)

```suggestion
          src={object.src?.kind === 'data'
```

mckdanoventa replied:
No, I went to bed thinking about this as well. 👍

wyozi commented on src/normalizer.ts:
```diff
@@ -54,7 +54,8 @@ export type InternalText = ObjectBase & {
   kind: "image";
-  url: string;
+  url?: string;
+  src?: InternalImageSrc;
```
Now I'm actually wondering why is this attribute optional at all 🤔

mckdanoventa replied:
Haha, you're right, it shouldn't be optional at all at this point.

wyozi commented on src/renderer.ts:
```diff
@@ -125,7 +129,13 @@ const renderSlide = async (
   if (node.backgroundColor) slide.background = { fill: node.backgroundColor };
-  if (node.backgroundImage) slide.background = { path: node.backgroundImage };
+  if (node.backgroundImage) {
+    if (typeof node.backgroundImage === 'string') {
```
This fallback will probably not be needed if we go for the full API breaking change

mckdanoventa replied:
👍

mckdanoventa commented on src/normalizer.ts:
```diff
@@ -54,7 +54,8 @@ export type InternalText = ObjectBase & {
 export type InternalImage = ObjectBase & {
   kind: "image";
-  url: string;
+  url?: string;
```
Ah, yes!

mckdanoventa commented on src/renderer.ts:
```diff
@@ -125,7 +129,9 @@ const renderSlide = async (
   slide.hidden = node.hidden;
   if (node.backgroundColor) slide.background = { fill: node.backgroundColor };
-  if (node.backgroundImage) slide.background = { path: node.backgroundImage };
+  if (node.backgroundImage) {
+      slide.background = { [node.backgroundImage.kind]: node.backgroundImage[node.backgroundImage.kind] };
+  }
```
If we want to keep the contract closer to the pptxgen, https://gitbrent.github.io/PptxGenJS/docs/usage-slide-options.html, we could also skip the `kind` property and propagate some changes like below: 

```suggestion
  if (node.backgroundImage) {
     slide.background =  node.backgroundImage['data' in node.backgroundImage ? 'data' : 'path']
  }
```

wyozi replied:
How would that work? I thought the pptxgen `background` property only accepts an object

mckdanoventa commented on src/renderer.ts:
```diff
@@ -125,7 +129,9 @@ const renderSlide = async (
   slide.hidden = node.hidden;
   if (node.backgroundColor) slide.background = { fill: node.backgroundColor };
-  if (node.backgroundImage) slide.background = { path: node.backgroundImage };
+  if (node.backgroundImage) {
+      slide.background = { [node.backgroundImage.kind]: node.backgroundImage[node.backgroundImage.kind] };
+  }
```
Ah, completely missed the object. Let me adjust a bit, our logic would end up looking something like: 

```suggestion
  if (node.backgroundImage) {
      const background = node.backgroundImage.entries[0]
      slide.background = { [background.key]: background.value };
  }
```

wyozi replied:
Ah, right. I think it's a bit better as the more explicit current version. Explicitly managing data tends to work better with typescript and is more readable. Also I think this has the risk of assigning `kind` to background

mckdanoventa replied:
Got it. With this iteration, I would only have the 

{ data: string } | { path: string} type for images. 

Anything else you'd like to see in this PR, @wyozi ?

wyozi replied:
Left one more comment. Also if you merge from master it should run tests etc

wyozi commented on src/normalizer.ts:
```diff
@@ -76,10 +76,22 @@ export type InternalShape = ObjectBase & {
+  data: string;
+}
+
+export type InternalImageSrc = (InternalPathSrc | InternalDataSrc ) & {
```
With `{kind: "data", data: string} | {kind: "path", path: string}` this will have a bit better type checking

mckdanoventa replied:
Cool, good to know, tx!

wyozi commented:
Cool. Thanks for the PR!
